### ISSUE TITLE
Convergence problems with GMG and spherical shells and large viscosity contrasts

### ISSUE THREAD

**Issue by gassmoeller** [geodynamics/aspect#4308]
@alarshi, @jdannberg and myself are experiencing convergence issues with the GMG preconditioner in spherical shell models with sharp viscosity variations at the boundary (AMG converges for the same models). I will continue to investigate, but wanted to open this issue to remind myself of our current understanding and to give @tjhei an update.

The attached model (and similar models) will sometimes randomly crash inside the preconditioner for the top left A block of the Stokes matrix. I have used the branch https://github.com/gassmoeller/aspect/tree/entropy_merged_output to output A and S block iterations in real time and it looks like this (full output attached):
```
*** Timestep 0:  t=0 years, dt=0 years
   Solving temperature system... 0 iterations.
   Solving Stokes system... Solving Stokes inner A block... 8 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 8 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 16 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 15 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 11 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 42 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 9 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 9 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 31 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 10 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 13 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 35 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 8 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... ---------------------------------------------------------
TimerOutput objects finalize timed values printed to the
screen by communicating over MPI in their destructors.
Since an exception is currently uncaught, this
synchronization (and subsequent output) will be skipped
to avoid a possible deadlock.
---------------------------------------------------------
```

Since it looks like the solver for the inner A block (CG at the moment) randomly has convergence issues in some outer iterations, I have tried replacing it with a FGMRES and a BiCGStab and both perform much better with less iterations and without convergence issues. Same model with FGMRES as inner solver:
```
*** Timestep 0:  t=0 years, dt=0 years
   Solving temperature system... 0 iterations.
   Solving Stokes system... Solving Stokes inner A block... 7 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 6 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 5 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 4 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 5 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 4 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 4 iterations.
Solving Stokes inner S block... 2 iterations.
Solving Stokes inner A block... 4 iterations.
```
this goes on for a while and eventually converges. This leads me to believe something is breaking the symmetry or positive definiteness of the A block in the matrix-free GMG solver. My current guess would be: 1. either the tangential boundary conditions (zero-slip bcs converge fine), but that could also just be a simpler solve, because the large viscosity contrasts at the boundary are the problematic regions), or 2. the projection of the viscosity to the different MG levels is introducing negative viscosities (it should not, viscosity in the material model is hard-capped to positive values between 1e17 and 5e22 and the projection should be limited to the maximum and minimum viscosity on the active mesh). I will continue to investigate, but if anyone has ideas I am open for suggestions.

[shell_simple_2d.prm.txt](https://github.com/geodynamics/aspect/files/6971794/shell_simple_2d.prm.txt)
[log.cg.txt](https://github.com/geodynamics/aspect/files/6971799/log.cg.txt)
[log.fgmres.txt](https://github.com/geodynamics/aspect/files/6971800/log.fgmres.txt)

**Comment by tjhei** [comment]
Great to have a small test problem to play with!

I appreciate your initial experiments as well, @gassmoeller .

I will take a look in the next few days. I have a few ideas...

**Comment by tjhei** [comment]
It converges without problems with 1,2,3, and 4 MPI ranks on the current master (a2a7ce93a) with deal.II 9.3.1 for me.

What is your deal.II version?

**Comment by tjhei** [comment]
The viscosity seems to be constant for me. Wrong file, @gassmoeller ?

**Comment by gassmoeller** [comment]
Sorry, I should have stated that you need to use the branch I have linked above. I hacked the material model to specify a viscosity (because I wanted a compressible model with a varying viscosity).

**Comment by gassmoeller** [comment]
Also: The branch has a bunch of other changes as well, but only the last two commits are relevant for this example. And I saw the same convergence problems with deal.II 9.2.0, 9.3.1, and 10.0.pre.

**Comment by tjhei** [comment]
Makes sense. I can reproduce it and I confirmed the issue: the rotational nullspace is the problem as I predicted in our video call a while ago. You can test this by doing zero velocity at the bottom.
Fixing this the right way will take a bit longer, but I have an idea for a workaround.

**Comment by tjhei** [comment]
matrix based:
```
   Solving Stokes system... 226+0 iterations.
   Solving Stokes system... 165+0 iterations.
   Solving Stokes system... 148+0 iterations.
```
matrix-free with my hack:
```
    Stokes: 208+0 iterations.
    Stokes: 156+0 iterations.
    Stokes: 143+0 iterations.
```
My branch is https://github.com/tjhei/aspect/tree/gmg_nullspace
I need to clean this up before I can open a PR but I can confirm that this test now works in 3d and 3d.
@alarshi do you want to test if this helps?

**Comment by tjhei** [comment]
matrix-based with new constraint (no big difference it seems):
```
   Solving Stokes system... 223+0 iterations.
   Solving Stokes system... 166+0 iterations.
   Solving Stokes system... 154+0 iterations.
```

**Comment by alarshi** [comment]
Hi @tjhei: Thank you for working on this! I have not tested your branch on our models yet, I will try that and update you.

**Comment by tjhei** [comment]
I just realized that I need to remove three rotational modes and not two in 3d. I will try that next.

**Comment by tjhei** [comment]
@alarshi I just pushed a new version to https://github.com/tjhei/aspect/tree/gmg_nullspace with an additional 3d constraint. Can you check if that makes a difference?

**Comment by elodie-kendall** [comment]
Hi Timo, I tested the GMG solver with my spherical shell models and it's massively helping reduce run time. However, I need to remove the null space to put back my free-slip boundaries. However, when I try to compile the branch you attach above I get the following attached errors. PS. is the GMG now also working a free surface (I saw you've been doing a PR for mesh deformation and the gmg)?
<img width="2358" alt="Screen Shot 2021-09-08 at 10 26 28" src="https://user-images.githubusercontent.com/75854106/132474145-b37ed9b8-d2a6-45cd-8fd0-3c2e59e1a46f.png">

**Comment by tjhei** [comment]
Elodie: the branch mentioned above is somewhat incomplete and I will prepare clean one we can merge soon (it contains other changes for @alarshi 's models). 
In the mean time, you can try adding ``#include <aspect/adiabatic_conditions/interface.h>`` to the simple.cc file with the error message.

> PS. is the GMG now also working a free surface

No, but this is the next step (maybe in the next few weeks). Right now we can only do other surface deformation. @zjiaqi2018 is working on it.

**Comment by elodie-kendall** [comment]
Ok, great thanks, I look forward to trying them soon :)

**Comment by elodie-kendall** [comment]
Any updates on using the GMG with null space removal please? Thanks so much!

**Comment by tjhei** [comment]
@elodie-kendall take a look at #4413

**Comment by elodie-kendall** [comment]
@tjhei It compiles and now the GMG solver works with null space removal. It's so much faster, that's amazing!! Thanks so much :)

**Comment by tjhei** [comment]
Great. I will need to do a couple more changes and I would hope you can test it again with your problem.

**Comment by elodie-kendall** [comment]
Sure, ping me when it's ready!

**Comment by lhy11009** [comment]
Hi, I want to comment on this issue, that I have a case running with block GMG in 3D cartesian geometry that failed to converge.
The same case with the Block AMG preconditioner doesn't have this issue. It seems to me that @tjhei's PR would address this issue. The only concern I have is that PR is not merged yet and is behind the current aspect. In this case, I am going to cherry-pick that PR onto the branch I am using. Meanwhile, I would like @tjhei to comment if that's the right thing to do if he has the time to do so.
BTW, I did an affinity test on stampede2 with the prm file I got from @gassmoeller, which concludes that using the "Block GMG" would reduce both the time for "Solving" and "Assembling" the Stokes by one order. Here are the figures for the summary.  Thus I am motivated to use this in my 3D model, which previously takes a considerable time to compute.
Block AMG:
![block AMG](https://user-images.githubusercontent.com/28665365/189498490-237e5c3d-19b1-4697-a5e3-a4f3c5b4bc59.png)
Block GMG:
![block GMG](https://user-images.githubusercontent.com/28665365/189498494-b5dbe92e-2085-4164-a080-a701ce4fd909.png)
I'd also like to comment to @mibillen and @rfildes that their issue with the "Block GMG" might also be addressed in this one and Timo's PR.

**Comment by tjhei** [comment]
> The only concern I have is that PR is not merged yet and is behind the current aspect.

Did you see #4413? It is current with the current main branch and is pretty close to being finished...

**Comment by tjhei** [comment]
> by one order.

This is consistent with our finding at https://onlinelibrary.wiley.com/doi/10.1002/nla.2375

**Comment by lhy11009** [comment]
Yes, I actually just tested out cherry-picking #4413 onto my branch and run the case again. However, the issue is not resolved. It could be:
1. There is a parameter I need to set for removing the nullspace. Is there one? I noticed the parameter.cc is changed but in the conversation below, it's mentioned that nothing needs to be changed.
2. I am having a different kind of issue. Here is the error message I am getting:

----------------------------------------------------
Exception 'ExcMessage (exception_message.str())' on rank 0 on processing:

--------------------------------------------------------
An error occurred in line <2779> of file </home1/06806/hyli/softwares/aspect/source/utilities.cc> in function
    void aspect::Utilities::linear_solver_failed(const string&, const string&, const std::vector<dealii::SolverControl>&, const std::exception&, const MPI_Comm&, const string&)
The violated condition was:
    false
Additional information:
    The iterative Stokes solver in
    StokesMatrixFreeHandlerImplementation::solve did not converge.

    The initial residual was: 1.368786e+14
    The final residual is: 8.547158e+13
    The required residual for convergence is: 3.712711e+13
    See output/solver_history.txt for the full convergence history.

    The solver reported the following error:

    --------------------------------------------------------
    An error occurred in line <2779> of file
    </home1/06806/hyli/softwares/aspect/source/utilities.cc> in function
    void aspect::Utilities::linear_solver_failed(const string&, const
    string&, const std::vector<dealii::SolverControl>&, const
    std::exception&, const MPI_Comm&, const string&)
    The violated condition was:
    false
    Additional information:
    The iterative (bottom right) solver in
    BlockSchurGMGPreconditioner::vmult did not converge.

    The initial residual was: 4.705162e-02
    The final residual is: 1.066801e-03
    The required residual for convergence is: 4.705162e-08

    The solver reported the following error:

    --------------------------------------------------------
    An error occurred in line <1280> of file
    </work2/06806/hyli/stampede2/Softwares/dealii/9.4.0/deal.II-v9.4.0/include/deal.II/lac/solver_cg.h>

    in function
    void dealii::SolverCG<VectorType>::solve(const MatrixType&,
    VectorType&, const VectorType&, const PreconditionerType&) [with
    MatrixType = aspect::MatrixFreeStokesOperators::MassMatrixOperator<3,
    1, double>; PreconditionerType = dealii::PreconditionMG<3,
    dealii::LinearAlgebra::distributed::Vector<double>,
    dealii::MGTransferMatrixFree<3, double> >; VectorType =
    dealii::LinearAlgebra::distributed::Vector<double>]
    The violated condition was:
                   solver_state == SolverControl::success
    Additional information:
    Iterative method reported convergence failure in step 100. The
    residual in the last step was 0.0010668.

    This error message can indicate that you have simply not allowed a
    sufficiently large number of iterations for your iterative solver to
    converge. This often happens when you increase the size of your
    problem. In such cases, the last residual will likely still be very
    small, and you can make the error go away by increasing the allowed
    number of iterations when setting up the SolverControl object that
    determines the maximal number of iterations you allow.

    The other situation where this error may occur is when your matrix is
    not invertible (e.g., your matrix has a null-space), or if you try to
    apply the wrong solver to a matrix (e.g., using CG for a matrix that
    is not symmetric or not positive definite). In these cases, the
    residual in the last iteration is likely going to be large.

**Comment by tjhei** [comment]
Do you have a simple .prm that you can share that shows the problem? If not, can you try to simplify?

**Comment by lhy11009** [comment]
Hi @tjhei. I haven't got a simplified version, but I have done some more tests. Here is what I found out.

**Comment by lhy11009** [comment]
First, the previous error message is from a branch with your PR #4413 cherry-picked. While I thought this doesn't solve the issue, I do notice the "The final residual" in the "BlockSchurGMGPreconditioner::vmult" changes from nan to 1.066801e-03. So that the PR  might help in some ways, possibly removing the issue caused by a null space.

**Comment by lhy11009** [comment]
Secondly, I was thinking of the viscosity contrast in the model. Previously, I have already limited the viscosity in [1e21, 1e24] by setting the " Minimum viscosity", and the "Maximum viscosity". I went further by  setting 
" Minimum viscosity" = "Maximum viscosity" = 1e21, by which the viscosity contrast is eliminated. This new case doesn't have any issues to converge. So the issue is indeed tied to the viscosity contrast somehow.

**Comment by lhy11009** [comment]
Thirdly, A suggestion from John is to test 
 "Use full A block as preconditioner"   = true
Which doesn't change the results, or the reported "final residual" in the "BlockSchurGMGPreconditioner::vmult".

**Comment by lhy11009** [comment]
Anyway, I'll try to get a simplified version of this. Here is the prm I have for now:

**Comment by lhy11009** [comment]
[test_ThD_gmg_mv1e21_peloton_case.prm.txt](https://github.com/geodynamics/aspect/files/9602831/test_ThD_gmg_mv1e21_peloton_case.prm.txt)

**Comment by lhy11009** [comment]
In short, it's a 3D subduction model in a box geometry, with newtonian rheology in the model.

**Comment by lhy11009** [comment]
Hi @tjhei , while turning this into a smaller test, I came up with the possibility that this could be a problem with the "Newton solver". It was hinted that the case I ran always crushed after "switching" from "defect correction" to "Newton". I assured this with a test I modified from the "Slab detachment model", where the viscosity has a strain-rate dependence. Here is a summary:
In the first case, I used "single Advection, iterated defect correction Stokes" as the solver scheme, and the run succeeded (prm - case_gmg_defected, stdout-task-496384.stdout). In the second case, I used "single Advection, iterated Newton Stokes" as the solver scheme and "block GMG" as the "Stokes solver type" and the run failed at time step 1 (prm - case_gmg_newton.prm, stdout - task-496395.stdout, stderr - task-496395.stderr).
[case_gmg_defected.prm.txt](https://github.com/geodynamics/aspect/files/9641720/case_gmg_defected.prm.txt)
[task-496384.stdout.txt](https://github.com/geodynamics/aspect/files/9641730/task-496384.stdout.txt)
[case_gmg_newton.prm.txt](https://github.com/geodynamics/aspect/files/9641727/case_gmg_newton.prm.txt)
[task-496395.stdout.txt](https://github.com/geodynamics/aspect/files/9641729/task-496395.stdout.txt)
[task-496395 .stderr.txt](https://github.com/geodynamics/aspect/files/9641728/task-496395.stderr.txt)

**Comment by tjhei** [comment]
Hi @lhy11009 : The last three prm files you posted are in a box geometry, so there is no nullspace in the system.

I saw that you used 
```
Linear solver tolerance = 1e-1
```
this might be problematic, I would at most use something like 1e-3.

**Comment by lhy11009** [comment]
I have modified the tests with "set Maximum linear Stokes solver tolerance = 1e-3". First, here are the results with the gmg and the Newton solver. This case still crashes at the end. The difference is the number of iterations in the Stokes system with each non-linear iteration, which increases from 1 to 4.
[case_gmg_newton_1e-3.prm.txt](https://github.com/geodynamics/aspect/files/9669664/case_gmg_newton_1e-3.prm.txt)
[task-498248.stdout.txt](https://github.com/geodynamics/aspect/files/9669674/task-498248.stdout.txt)
[task-498248.stderr.txt](https://github.com/geodynamics/aspect/files/9669676/task-498248.stderr.txt)

**Comment by lhy11009** [comment]
I went further to set this to "1e-6". This would increase the number of interactions in the Stokes system to about 13 with each non-linear iteration. But the case still crushes upon switching to the Newton solver.
[case_gmg_newton_1e-6.prm.txt](https://github.com/geodynamics/aspect/files/9669772/case_gmg_newton_1e-6.prm.txt)
[task-498361.stdout.txt](https://github.com/geodynamics/aspect/files/9669774/task-498361.stdout.txt)
[task-498361.stderr.txt](https://github.com/geodynamics/aspect/files/9669783/task-498361.stderr.txt)

**Comment by lhy11009** [comment]
For comparison, I also tested the "block AMG". This case ran successfully.
[case_amg_newton_1e-3.prm.txt](https://github.com/geodynamics/aspect/files/9669793/case_amg_newton_1e-3.prm.txt)
[task-498251.stdout.txt](https://github.com/geodynamics/aspect/files/9669794/task-498251.stdout.txt)

**Comment by lhy11009** [comment]
The Fourth test is "block GMG" with the defect correction Picard solver. This also ran successfully.
[case_gmg_picard_1e-3.prm.txt](https://github.com/geodynamics/aspect/files/9669809/case_gmg_picard_1e-3.prm.txt)
[task-498254.stdout.txt](https://github.com/geodynamics/aspect/files/9669814/task-498254.stdout.txt)

**Comment by lhy11009** [comment]
The takeaway is still that the issue is with the combination of the "block GMG" with "iterated Newton Stokes"

**Comment by anne-glerum** [comment]
@lhy11009 is possibly related to #4563, although in this issue the DC Picard solver seems to work.
